File: jobs/Backend/src/Core.Infra/Dtos/ExchangeRateDto.cs
namespace Core.Infra.Dtos
{
    public class ExchangeRateDto
    {
        public ExchangeRateDto(string country, string currency, string baseAmount, string code, string rate)
        {
            Country = country;
            Currency = currency;
            BaseAmount = baseAmount;
            Code = code;
            Rate = rate;
        }

        public string Country { get; }
        public string Currency { get; }
        public string BaseAmount { get; }
        public string Code { get; }
        public string Rate { get; }
    }
}

File: app/dtos/exchange_rate_dto.rb
module DTOs
  class ExchangeRateDTO
    attr_reader :from, :to, :rate, :date

    def initialize(from, to, rate, date = nil)
      @from = from
      @to = to
      @rate = rate
      @date = date
    end

    # Create DTO from domain model
    def self.from_domain(exchange_rate)
      new(
        exchange_rate.from.code,
        exchange_rate.to.code,
        exchange_rate.rate,
        exchange_rate.respond_to?(:date) ? exchange_rate.date : nil
      )
    end

    # Create a collection of DTOs from domain models
    def self.from_domain_collection(exchange_rates)
      exchange_rates.map { |rate| from_domain(rate) }
    end

    def to_h
      {
        from: from,
        to: to,
        rate: rate,
        date: date
      }
    end
  end
end